File: CamStudy/CamStudy/ViewModels/CreateStudyRoomViewModel.swift
import Foundation

@MainActor
final class CreateStudyRoomViewModel: ObservableObject {
    @Published var name = ""
    @Published var maxParticipants = 6
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var createdRoom: StudyRoom?

    private let firestoreService = FirestoreService.shared

    func createRoom(hostId: String, hostNickname: String) {
        guard !name.trimmingCharacters(in: .whitespaces).isEmpty else {
            errorMessage = "방 이름을 입력해주세요."
            return
        }
        isLoading = true
        errorMessage = nil

        let room = StudyRoom(
            id: UUID().uuidString,
            name: name,
            hostId: hostId,
            hostNickname: hostNickname,
            maxParticipants: maxParticipants,
            participantIds: [hostId],
            participantNicknames: [hostId: hostNickname],
            createdAt: Date(),
            isActive: true
        )

        firestoreService.createStudyRoom(room) { [weak self] result in
            Task { @MainActor in
                guard let self else { return }
                self.isLoading = false
                switch result {
                case .success:
                    self.createdRoom = room
                case .failure(let error):
                    self.errorMessage = error.localizedDescription
                }
            }
        }
    }
}

File: camstudy_app/lib/viewmodels/create_study_room_view_model.dart
import 'package:flutter/foundation.dart';

import '../models/study_room.dart';
import '../services/firestore_service.dart';

class CreateStudyRoomViewModel extends ChangeNotifier {
  String name = '';
  int maxParticipants = 6;
  bool isLoading = false;
  String? errorMessage;
  StudyRoom? createdRoom;

  final _firestoreService = FirestoreService.shared;

  void setMaxParticipants(int value) {
    maxParticipants = value;
    notifyListeners();
  }

  Future<void> createRoom({
    required String hostId,
    required String hostNickname,
  }) async {
    if (name.trim().isEmpty) {
      errorMessage = '방 이름을 입력해주세요.';
      notifyListeners();
      return;
    }
    isLoading = true;
    errorMessage = null;
    notifyListeners();

    try {
      createdRoom = await _firestoreService.createStudyRoom(
        name: name,
        hostId: hostId,
        hostNickname: hostNickname,
        maxParticipants: maxParticipants,
      );
    } catch (e) {
      errorMessage = e.toString();
    }
    isLoading = false;
    notifyListeners();
  }
}
